plugins/plugin_utils/keys_filter.py

```diff
@@ -93,16 +93,43 @@ def _keys_filter_target_str(target, matching_parameter):
 
 def _keys_filter_target_dict(target, matching_parameter):
     """test:
-    * target is a list of dictionaries
-    * ...
+       * target is a list of dictionaries with attributes 'after' and 'before'.
+       * If matching_parameter=regex the attributes 'before' must be valid regex.
+       * Otherwise, the attributes 'before' must be strings.
+       convert and return:
+       * Dictionary of attributes 'before' and 'after'
+       * Dictionary of compiled regex of attributes 'before' and 'after'
     """
 
-    # TODO: Complete and use this in filter replace_keys
+    if not isinstance(target, list):
+        msg = "The target must be a list. target is %s."
+        raise AnsibleFilterError(msg % (target, type(target)))
 
-    if isinstance(target, list):
-        for elem in target:
-            if not isinstance(elem, Mapping):
-                msg = "The target items must be dictionary. %s is %s"
-                raise AnsibleFilterError(msg % (elem, type(elem)))
+    for elem in target:
+        if not isinstance(elem, Mapping):
+            msg = "The target items must be dictionaries. %s is %s"
+            raise AnsibleFilterError(msg % (elem, type(elem)))
+        if not all(k in elem for k in ('before', 'after')):
+            msg = "All dictionaries in target must include: after, before."
+            raise AnsibleFilterError(msg)
 
-    return
+    before = [d['before'] for d in target]
+    after = [d['after'] for d in target]
+
+    if matching_parameter == 'regex':
+        try:
+            tt = map(re.compile, before)
+            td = dict(zip(tt, after))
+        except re.error:
+            msg = ("The attributes before must be valid regex if matching_parameter=regex."
+                   " Not all items are valid regex in: %s")
+            raise AnsibleFilterError(msg % before)
+    else:
+        if not all(isinstance(item, string_types) for item in before):
+            msg = ("The attributes before must be strings."
+                   " Not all items are strings in: %s")
+            raise AnsibleFilterError(msg % before)
+        else:
+            td = dict(zip(before, after))
+
```

Comment:
You're right. There are unnecessary problems using dict. I switched the code to 
list and implemented all your suggestions.  Would it be possible to review the changes?

There are changes in the interface `_keys_filter_params(data, matching_parameter)` because `target` is not tested in this function anymore. I updated and tested also `keep_keys` and `remove_keys`. 

Thank you!